shared/middleware/assets_api.rb

```diff
@@ -4,60 +4,81 @@
 
 class AssetsApi < Sinatra::Base
 
-  # Only allow specific image and sound types to be uploaded by users.
-  ALLOWED_FILE_TYPES = %w(.jpg .jpeg .gif .png .mp3)
+  def get_impl(endpoint)
+    case endpoint
+    when 'assets'
+      AssetBucket
+    when 'sources'
+      SourceBucket
+    else
+      not_found
+    end
+  end
+
+  def allowed_file_type?(endpoint, extension)
+    case endpoint
+    when 'assets'
+      # Only allow specific image and sound types to be uploaded by users.
+      %w(.jpg .jpeg .gif .png .mp3).include? extension
+    when 'sources'
+      # Only allow JavaScript and Blockly XML source files.
+      %w(.js .xml).include? extension
+    else
+      not_found
+    end
+  end
 
   helpers do
-    %w(core.rb asset_bucket.rb storage_id.rb).each do |file|
+    %w(core.rb bucket_helper.rb asset_bucket.rb source_bucket.rb storage_id.rb).each do |file|
       load(CDO.dir('shared', 'middleware', 'helpers', file))
     end
   end
 
   #
-  # GET /v3/assets/<channel-id>
+  # GET /v3/(assets|sources)/<channel-id>
   #
   # List filenames and sizes.
   #
-  get %r{/v3/assets/([^/]+)$} do |encrypted_channel_id|
+  get %r{/v3/(assets|sources)/([^/]+)$} do |endpoint, encrypted_channel_id|
     dont_cache
     content_type :json
 
-    AssetBucket.new.list(encrypted_channel_id).to_json
+    get_impl(endpoint).new.list(encrypted_channel_id).to_json
```

Comment:
how about renaming get_impl to bucket_helper (or maybe bucket_impl) to give a little more context?